docs/generic-templates.md

```diff
@@ -0,0 +1,173 @@
+# Generic templates
+
+This presentation ingores a few things which are easy to implement but would add unnecessary complexity to the presentation:
+- `observer`, `ensure` and `agreement` clauses in templates
+- `preconsuming`, `postconsuming` and `nonconsuming` declarations
+- the implicit `Archive` choice
+- contract keys
+
+We also assume that DAML-LF provides a function to coerce contract ids, which we expose as
+```haskell
+coerceContractId : ContractId a -> ContractId b
+coerceContractId = primitive @"BECoerceContractId"
+```
+
+## Surface syntax
+
+A generic template can be defined using the following syntax:
+```haskell
+template D a => T a with
+    f_1 : U_1; ...
+  where
+    signatory sign
+
+    choice C : R with g_1 : V_1; ...
+      controller ctrl
+      do act
+```
+This generic template has one type parameter `a`, which is subject to the constraint `D a`. This constraint can be empty, written as
+```haskell
+template T a with ...
+````
+All types `R`, `U_i` and `V_j` have the type variable `a` in scope.
+
+An instantiation of this generic template with an actual type `S` in place of type parameter `a` can be declared through
+```haskell
+template instance I = T S
+```
+The `I` is the name of the underlying DAML-LF type and template that will be created for this instantiation. This type will also exist in DAML but is not intended to be used there.
+
+Extending this syntax to more than one type parameter is trivial.
+
+## Underlying type classes
+
+We split the current `Template` type class into three type classes serving the following purpose:
+1. `LfTemplate` indicates an "actual template" which will cause a template definition in DAML-LF. Our example above will yield `instance LfTemplate I`. The `LfTemplate` type class serves the same purpose as the current `Template` type class but is not intended to be used in DAML. Instances of `LfTemplate` are generated for generic template instantiations and non-generic template definitions.
+1. `Template` indicates a "logical template" which exists only in DAML but not in DAML-LF. Our example above will yield `instance D a => Template (T a)`. Instances of `Template` are generated for all template definitions, generic and non-generic ones.
+1. `Creatable` links an instance of `Template` with an instance of `LfTemplate` (at the value level). Our example above will yield `instance Creatable (T S)`. Instances of `Creatable` are generated for generic template instantiations and non-generic template definitions.
```

Comment:
Don't like the name Createable

Reply:
We could rename `Template` to `PreTemplate` and `Createable` to `Template`. But let's_ do the bike shedding later. Maybe the split between these two classes will disappear during the evolution of this design.